### DISCUSSION TITLE
请求一个最简单的 v5 DNS 配置，用于测试

### DISCUSSION THREAD

**Discussion by wy16W2pIilK1xgqN** [v2fly/v2ray-core#2341] [Q&A] [answered]
一直用v2ray v5启动 v4 版配置，一直工作良好但总提示DNS配置过时，
日前想切换到v5版的DNS 配置，修改后发现 DNS 查询一直失败，什么都解析不出来，
这两天一直在研究死活不成功，希望大家能给个最简单的v5版等效配置，让我测试一下。
只要求 内置 DNS 能返回解析地址就行，"fakedns" 可不配置。

目前v4 配置， 局域网客户端，通过  "FakeDns-In"，查询DNS

```
........
    "inbounds": [
        {
            "port": 13253,
            "protocol": "dokodemo-door",
            "tag": "FakeDns-In",
            "settings": {
                "address": "8.8.8.8",
                "port": 53,
                "network": "udp"
            }
        }，
                {
            "protocol": "dns",
            "tag": "Dns-Out",
            "streamSettings": {
                "sockopt": {
                    "mark": 255
                }
            }
        }
    ]
........
    // 截取DNS
    {
        "type": "field",
        "outboundTag": "Dns-Out",
        "inboundTag": [
            "Socks-In",
            "Http-In",
            "FakeDns-In"
        ],
        "network": "udp",
        "port": 53
    }
........
    // DNS 配置
    "dns": {
        "servers": [
            "fakedns",
            "8.8.8.8",
            "1.1.1.1",
            "localhost"
        ]
    },
    "fakedns": {
        "ipPool": "172.20.0.0/16",
        "poolSize": 65534
    }
........


```

**Comment by mydogshitgold** [comment]
The exact conversion to v5 format is
```json
{
  "inbounds": [
    {
      "tag": "FakeDns-In",
      "protocol": "dokodemo-door",
      "port": 13253,
      "settings": {
        "address": "8.8.8.8",
        "port": 53,
        "networks": "udp"
      }
    }
  ],
  "outbounds": [
    {
      "tag": "Dns-Out",
      "protocol": "dns",
      "streamSettings": {
        "socketSettings": {
          "mark": 255
        }
      }
    }
  ],
  "router": {
    "rule": [
      {
        "tag": "Dns-Out",
        "inboundTag": [
          "Socks-In",
          "Http-In",
          "FakeDns-In"
        ],
        "networks": "UDP",
        "portList": "53"
      }
    ]
  },
  "dns": {
    "nameServer": [
      {
        "address": {
          "address": "fakedns"
        },
        "fakeDns": {
          "pools": [
            {
              "ipPool": "172.20.0.0/16",
              "lruSize": 65534
            }
          ]
        }
      },
      {
        "address": {
          "address": "8.8.8.8"
        }
      },
      {
        "address": {
          "address": "1.1.1.1"
        }
      },
      {
        "address": {
          "address": "localhost"
        }
      }
    ]
  }
}
```
### NOTE
- Address block 172.16.0.0/12 is used for local communications within a private network, just like 192.168.0.0/16. Using it as FakeDNS IP pool may cause addressing conflicts. It is safer to use address block 198.18.0.0/15, which is originally used for benchmarking.
- When using HTTP proxy, name resolution is done on the server side.

**Comment by wy16W2pIilK1xgqN** [comment]
谢谢你能来帮助我💐
但情况一样，我依然无法得到DNS 响应。😒

**Reply by mydogshitgold** [reply]
I suspect that there may be flaws in your workflow, If this exact conversion of the configuration you have provided does not work. 

Do you mind provide more details? e.g.
- full configuration
- operating system
- iptables
- routing table

**Reply by wy16W2pIilK1xgqN** [reply]
我是通过 WIn 11，在本机运行  v2ray 5.4 ，
为了防止其它设置的干扰，也在全新的 win 11 虚拟机中重复了相同的测试。 
基于测试环境，我认为与路由表和其他网络设置关系不大。


测试命令  `doggo.exe @udp://127.0.0.1:13253  qq.com`

这是用来测试的两个完整配置文件。
PS：文件名称请去掉 '.txt'
[v4.json.txt](https://github.com/v2fly/v2ray-core/files/10790153/v4.json.txt)
[v5.json.txt](https://github.com/v2fly/v2ray-core/files/10790164/v5.json.txt)

**Reply by mydogshitgold** [reply]
> Based on the test environment, I don't think it has much to do with routing tables and other network settings.

It looks like that you are using transparent proxy (since you are using SO_MARK, however, please note that `sockopt.mark`[ is **NOT** supported on Windows](https://www.v2fly.org/config/transport.html#sockoptobject)), then routing tables and firewall does do important roles as they are used in combination to route all traffics from you PC to V2Ray.

It's unusual to use a non-standard port :13253 for DNS, most of software and libraries do not support anything but port :53 for DNS. And since you are already using transparent proxy, why not just  hijack DNS traffics using firewall and V2Ray routing rules? You could just write
```json
{
  "outboundTag": "Dns-Out",
  "inboundTag": [
    "Transparent-In"
  ],
  "port": 53,
  "network": "tcp,udp",
  "type": "field"
}
```
Again, this is where firewall rules are needed. This is an example of hijacking DNS traffics on LAN:
```python
-A proxy -d 192.168.0.0/16 -p udp ! --dport 53 -j RETURN
-A proxy -d 192.168.0.0/16 -p tcp ! --dport 53 -j RETURN
```
---
What is `routing.strategy`? AFAIK V2Ray does not have this configuration.

Where is `Transparent-In` inbound? It is not in the configurations you provided. Are you sure they are full configuration?

Also, Windows is not a good choice to implement transparent proxy, as there is no clear documentation and very few resources on the internet. Using a Linux device or virtual machine as network gateway for windows is easier to implement and more mature.

If you have to use Windows to achieve effects similar to transparent proxy, then you can tunnel [Wireguard](https://en.wikipedia.org/wiki/Wireguard) over v2ray. If you don't have control over your proxy server, you can try some free Wireguard provider, such as (proton.me)[https://proton.me] or [Cloudflare WARP via wgcf](https://github.com/ViRb3/wgcf), both of them offers unlimited bandwidth.

- [Wireguard official website](https://www.wireguard.com)
- [Wireguard tutorial](https://www.wireguard.com/quickstart/)

**Reply by mydogshitgold** [reply]
Most of v5 configuration you provided is malformed.

- There is no `routing` in v5 configuration format, it's called `router`.
- `routing.balancers` should be `router.balancingRule`
- `routing.rules` should be `router.rule`
- `network` field in routing rule should be either `["TCP", "UDP"]` or `"TCP"|"UDP"`, not "tcp,udp"
- remove `"type": "field"`, they are not in v5.
- `outboundTag` field in routing rule should be simply `tag`
- `port` field in routing rule should be `portList`, and it's a string.
- and so on

You can refer to [documentation](https://www.v2fly.org/v5/config/overview.html) and use this [JSON schema](https://cdn.jsdelivr.net/gh/EHfive/v2ray-jsonschema/v5-config.schema.json) to correct your configuration.

**Reply by wy16W2pIilK1xgqN** [reply]
我找到问题了，感谢你的帮助。🎉🎉

我不懂英文，您显然也不擅长中文，
如此艰难的沟通，你能一直帮助我，
非常感激你，远方的朋友祝你合家安康。
╰(*°▽°*)╯

**Comment by wy16W2pIilK1xgqN** [comment, marked as answer]
问题归纳，已提醒其他人

由 @mydogshitgold 的解答，请感谢他

问题原因：混用V 4 和 v5 配置。
解决方法： 不要混用配置。

 


v5配置测试命令
v2ray.exe  run  -c  "配置文件路径" -format jsonv5

 正确的配置 
```
{
  "inbounds": [
    {
      "tag": "FakeDns-In",
      "protocol": "dokodemo-door",
      "port": 13253,
      "settings": {
        "address": "114.114.114.114",
        "port": 53,
        "networks": "udp"
      }
    }
  ],
  "outbounds": [
    {
      "tag": "Dns-Out",
      "protocol": "dns"
    },
    {
      "protocol": "freedom",
      "tag": "Direct"
    }
  ],
  "router": {
    "rule": [
      {
        "tag": "Dns-Out",
        "inboundTag": [
          "FakeDns-In"
        ],
        "networks": "UDP",
        "portList": "53"
      },
      {
        "tag": "Direct",
        "networks": "tcp,udp"
      }
    ]
  },
  "dns": {
    "nameServer": [
      {
        "address": {
          "address": "fakedns"
        },
        "fakeDns": {
          "pools": [
            {
              "ipPool": "172.20.0.0/16",
              "lruSize": 65534
            }
          ]
        }
      },
      {
        "address": {
          "address": "114.114.114.114"
        }
      },
      {
        "address": {
          "address": "1.1.1.1"
        }
      },
      {
        "address": {
          "address": "localhost"
        }
      }
    ],
    "tag": "v2-dns"
  }
}


```
